julia 1.10.1

korbinian90 opened the pull request:
@stebo85 
This PR uses the official julia docker image

A difference is that it directly drops the user into a running julia REPL (when testing it after an offline build, not sure if this happens in Neurodesk as well)

Another question about using julia, since there are multiple common ways to use julia:
1) REPL
2) vscode as IDE with julia extension
- vscode executes files or line-by-line in an interactive julia REPL in the vscode terminal. The path to the julia executable can be given to vscode
3) Jupyter or Pluto notebook

Can we create user-friendly ways to run julia for all these scenarios?

1) REPL is easy
2) I tried two things for vscode
- Opening the julia container and running vscode. This does not work as the command "code" is not found
- Starting vscode and setting the Executable Path for julia to "/cvmfs/neurodesk.ardc.edu.au/containers/julia_1.4.1_20210105/julia". This worked, but needs good documentation so that users can follow it.

stebo85 reviewed (changes requested):
Dear @korbinian90,

That's great thank you :)

the Julia base image could work and makes the install script a bit easier to read. If you want to drop users into a Julia REPL on startup you need the entrypoint setup:
```
--entrypoint /REPLACE_WITH_PATH_TO_JULIA_BINARY \ 
```

To get VScode to work you need to install VS code in the same container, e.g. following this example: https://github.com/NeuroDesk/neurocontainers/blob/master/recipes/code/build.sh
-> There we also have Julia added to the container, so you can test this setup already in the general code container.

The last thing would be to add a Julia kernel to the "global" jupyter environment. If you could play with this that would be amazing! We somehow would need to expose the julia kernel to the jupyter system. I played a bit with this using adding custom kernels: https://www.neurodesk.org/tutorials-examples/tutorials/programming/conda/ - but it's not perfect yet. Instead of installing the kernels I would love to "load" the kernels from existing containers.

korbinian90 commented:
We could have 2 or 3 start options for julia, with vscode, repl and the container. Not sure if all are necessary.

```
"julia": {
        "apps": {
            "julia 1.4.1": {
                "version": "20210105",
                "exec": ""
            },
            "1.10.1": {
                "version": "2024...",
                "exec": ""
            },
            "1.10.1 vscodeGUI": {
                "version": "2024...",
                "exec": "code"
            },
            "1.10.1 REPL": {
                "version": "2024...",
                "exec": "julia"
            },
        },
        "categories": ["programming"]
    }
```
Or would it be better to use entrypoint instead of this apps configuration?

I tried the vscode container, julia seems to work there. The only issue might be that vscode extension cannot be updated or installed after the container is created.

stebo85 commented:
I think this configuration is more flexible and looks good :) the entrypoint makes things a bit inflexible sometimes

korbinian90 commented:
I took the code container as base and removed the go and python stuff. The vscode extensions are frozen, but I think this is good enough.

I'm not sure, maybe the module.sh or other things are unnecessary here.

stebo85 reviewed (approved):
great - that could work :)

stebo85 commented:
@korbinian90 - can you test? code ist not yet working - I think it needs a bindmount of /run but I don't understand why my old container was working correctly yet.

korbinian90 commented:
@stebo85 yes, the same issue, when I tried with `bash /neurocommand/local/fetch_and_run.sh julia 1.10.1 20240227`
code doesn't open. `/usr/bin/code -v` shows the version number, so it is installed
I noticed that code doesn't work in the docker container when trying it locally before pushing it, but I thought maybe GUIs don't work there.

I removed these dependencies from the installation when adapting from the code recipe to julia: cryptsetup squashfs-tools lua-bit32 lua-filesystem lua-json lua-lpeg lua-posix lua-term lua5.2 imagemagick python-pyqt5 midori
They seemed to be unrelated to vscode and "midori" couldn't be found for installation
Otherwise I changed debian:9 to debian

korbinian90 commented:
I tried building the original code container, it failed to build.

After adjusting debian:9 to debian and removing midori and python-pyqt5, the container was building, but at least locally the same problem still existed.

stebo85 commented:
ok, good to know. I think it needs the /run directory bind mounted for vscode to work. I have an idea how to do this and will test that next week.

stebo85 commented:
I tried to mount the /run directory but it doesn't work yet, because dbus being missing

stebo85 commented:
Dear @korbinian90 - @Vbitz found the problem! VScode needs libX11-xcb - I now added the library to the container and it's working now :) It's currently in the release process, so should be ready to test in a few hours.

stebo85 commented:
ok, do you have an idea why that happens? I have a look at it next week.

korbinian90 commented:
@stebo85 I managed to get a screenshot of the error
It seems to be attaching REPL and vscode to the container name and looking for a non-existing container

Starting vscodeGUI:
![image](https://github.com/NeuroDesk/neurocontainers/assets/1307522/7dc4e051-3b1c-45db-9968-facff7ade662)

Starting REPLGUI:
![image](https://github.com/NeuroDesk/neurocontainers/assets/1307522/612c5c35-de56-4e5e-aa2f-d12c5b2ce02d)

stebo85 commented:
ahh, found the problem! The menu items were incorrectly named. I just fixed that now. Can you click the Update button in the menu and then it works (It will create two entries for now, but the next release will have this appropriately fixed)

stebo85 commented:
Great :) The new Neurodesk release 2024-03-27 has this fix included now.
